docs/okeps/okep-5224-connecting-udns/okep-5224-connecting-udns.md

```diff
@@ -0,0 +1,543 @@
+networkSelectionType: ClusterUserDefinedNetworks
+clusterUserDefinedNetworkSelector:
+  networkSelector:
+    matchExpressions:
+    - key: kubernetes.io/metadata.name
+      operator: In
+      values:
+      - blue-network
+      - green-network
+networkSelectionType: PrimaryUserDefinedNetworks
+primaryUserDefinedNetworkSelector:
+  namespaceSelector:
+    matchExpressions:
+    - key: kubernetes.io/metadata.name
+      operator: In
+      values:
+      - yellow
+      - red
+connectSubnets: # can have at most 1 CIDR for each family type
+  - cidr: 192.168.0.0/10
+    # This slice must be set to the maximum number of networks
+    # a user plans to connect together (similar to the podsubnet
+    # slice that equates max number of pods on each node). This
+    # slice is then handed out to each node.
+    # Can be removed if we go with Idea2 in the implementation section
+    hostSubnet: 24
+ featuresEnabled:
+   - podNetworkConnectivity
+     # enabled through a well-known label on service if we need finer-grained control
+   - clusterIPServicesConnectivity
+   - networkPolicyEnforcement
+```
+* `featuresEnabled` field will ensure we can add support for enabling more types of
+  features for the connected UDNs granularly in the future. At least one of these
+  options must be set and validations will be added for which combinations of these
+  options could/should be set together
+* If `podNetworkConnectivity` is enabled then admin is asking for full pod2pod
+  connectivity across UDNs
+* If `clusterIPServicesConnectivity` is enabled then admin is asking for
+  clusterIPs to be reachable across UDNs.
+    * If we don’t want to allow ALL clusterIPs to be exposed to other UDNs,
+      then on a per-service level in addition to the UDN CR requesting for it,
+      we can add a granularity level of “making that service be a candidate for UDNConnectivity”
+* If `networkPolicyEnforcement` is enabled then network policies will span across UDNs.
+   * If we want to only allow some policies then again label based granularity will be needed.
+* `connectSubnet` field is used to take a configurable subnet from the end user that will
+  be used to connect the different networks together. The slice needs to be big enough to
+  accommodate an IP per node per network. So if there are N nodes in your cluster and M
+  networks you should have N*M IPs for allocation in this slice. This value is expected
+  to be configured uniquely across multiple UDNConnect CRs that select the same network
+  i.e. If a given network is selected as part of more than 1 connect API, each of those
+  connect APIs must have a unique subnet. If a user accidentally configures overlapping
+  values, we must have a status check to block and report this in the status (details to be fleshed out).
+
+#### API Validations
+
+
+#### Valid ways of Connecting/Selecting UDNs
+
+TBD
+* A network could be selected as part of more than one network, so what is the expectation
```

Comment:
perhaps we should add a status field on the UDN to show which connect CRD name it is tied to. Maybe also a finalizer that wont let the UDN be deleted until the connect CRD is deleted. Then if someone creates another connect CRD it will fail because the UDN already is connected by another CRD.

I guess though, what if we have A connected to B, and now someone makes a CRD to connect B to C. Do we want to support that? Or should the original CRD be updated so that it says A,B,C are all connected.